src/libstd/sys/unix/condvar.rs

```diff
@@ -100,10 +107,60 @@ impl Condvar {
     // https://github.com/llvm-mirror/libcxx/blob/release_35/src/condition_variable.cpp#L46
     // https://github.com/llvm-mirror/libcxx/blob/release_35/include/__mutex_base#L367
     #[cfg(any(target_os = "macos", target_os = "ios", target_os = "android"))]
-    pub unsafe fn wait_timeout(&self, mutex: &Mutex, dur: Duration) -> bool {
+    pub unsafe fn wait_timeout(&self, mutex: &Mutex, mut dur: Duration) -> bool {
         use ptr;
         use time::Instant;
 
+        // 1000 years
+        let max_dur = Duration::from_secs(1000 * 365 * 86400);
+
+        if dur > max_dur {
+            // Apparently `tv_sec` overflows somewhere around
+            // now + 0x100_0000_0000_0000 seconds in `pthread_cond_timedwait`
+            // in macOS Sierra, which results in immediate return
+            // of `pthread_cond_timedwait`.
+            //
+            // However, to avoid catching possible bugs of other OSes, timeout is
+            // round down to 1000 years instead of 3000000 years.
+            //
+            // This is valid, because `wait` is allowed to wake up spuriously.
+            //
+            // This program demonstrates the issue:
+            //
+            // ```
+            // #![feature(libc)]
+            // extern crate libc;
+            //
+            // use std::ptr;
+            //
+            // fn main() {
+            //     unsafe {
+            //         let mut mutex = libc::PTHREAD_MUTEX_INITIALIZER;
+            //         let mut condvar = libc::PTHREAD_COND_INITIALIZER;
+            //         let r = libc::pthread_mutex_lock(&mut mutex);
+            //         assert_eq!(r, 0);
+            //
+            //         let now = libc::time(ptr::null_mut());
+            //         let timeout = libc::timespec {
+            //             // this version properly waits
+            //             //tv_sec: now + 0x100_0000_0000_0000 - 0x100_0000,
+            //             // this exits immediately
+            //             tv_sec: now + 0x100_0000_0000_0000,
+            //             tv_nsec: 0,
+            //         };
+            //
+            //         println!("timespec {} {}", timeout.tv_sec, timeout.tv_nsec);
+            //
+            //         let r = libc::pthread_cond_timedwait(&mut condvar, &mut mutex, &timeout);
+            //         assert!(r == libc::ETIMEDOUT || r == 0);
+            //         println!("$");
+            //     }
+            // }
+            // ```
```

Comment:
This is a pretty long comment to work with, could the program be extracted to something like a gist? Could the example program be C and not Rust?